harness/determined/pytorch/trainer.py

```diff
@@ -0,0 +1,285 @@
+    ):
+        self.trainer = trainer
+        self.step_type = TrainStepType.EPOCH if max_epochs else TrainStepType.BATCH
+        self.max_steps = max_epochs if self.step_type == TrainStepType.EPOCH else max_batches
+        self.batches_trained = 0
+        self.epochs_trained = 0
+        self.steps_completed = 0
+        self.checkpoint_period = min_checkpoint_period
+        self.validation_period = min_validation_period
+        self.reporting_period = min(min_checkpoint_period, min_validation_period)
+        self.training_metrics = []
+
+    def _steps_remaining(self):
+        return self.max_steps - self.steps_completed
+
+    def run(self):
+        while self._steps_remaining() > 0:
+            self.trainer.on_training_epoch_start(self.epochs_trained)
+            for batch_idx, batch in enumerate(self.trainer.train_loader):
+                self._train_batch(batch, self.epochs_trained, batch_idx)
+                if not self._steps_remaining() > 0:
+                    return
+            self.trainer.on_training_epoch_end(self.epochs_trained)
+            self._step_epoch()
+
+    def _train_batch(self, batch, epochs, batch_idx):
+        self.trainer.context._current_batch_idx = self.batches_trained
+        batch_metrics = self.trainer.trial.train_batch(batch, epochs, batch_idx)
+        self.training_metrics.append(batch_metrics)
+        self._step_batch()
+
+    def _train_step(self):
+        self.steps_completed += 1
+        if self.steps_completed % self.reporting_period == 0:
+            self.trainer.report_training_metrics(self.steps_completed, self.training_metrics)
+            self.training_metrics = []
+        if self.steps_completed % self.validation_period == 0:
+            val_metrics = self.trainer.validate()
+            self.trainer.report_validation_metrics(self.steps_completed, val_metrics)
+
+    def _step_batch(self):
+        self.batches_trained += 1
+        if self.step_type == TrainStepType.BATCH:
+            self._train_step()
+
+    def _step_epoch(self):
+        self.epochs_trained += 1
+        if self.step_type == TrainStepType.EPOCH:
+            self._train_step()
+
+
+def initialize_distributed_backend():
+    distributed_backend = det._DistributedBackend()
+    if distributed_backend.use_horovod():
+        hvd.require_horovod_type("torch", "PyTorchTrial is in use.")
+        hvd.init()
+        return core.DistributedContext.from_horovod(horovod.hvd)
+    elif distributed_backend.use_torch():
+        if torch.cuda.is_available():
+            dist.init_process_group(backend="nccl")  # type: ignore
+        else:
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+    else:
+        print(f"Backend {distributed_backend} not supported")
+
+
+@contextlib.contextmanager
+def init():
+    distributed_context = initialize_distributed_backend()
+    core_context = core.init(distributed=distributed_context)
```

Comment:
this needs to be `with core.init(...) as core_context`.

The way it's written right now, `core_context.close()` doesn't get ever called.

Reply:
I'd say at least 90% of the time, a contextmanager that doesn't have a `try: yield; finally: ...` or a `with X: yield` is not doing what you want, because if you are using a contextmanager it is so that you can do some cleanup step after execution continues from `yield`.

There's like 10% of the time where the steps you would complete afterwards should only be completed in success cases, and that's the only time you have neither a `try` nor a `with`.